### ISSUE TITLE
Marathon healthchecks on wrong address

### ISSUE THREAD

**Issue by scalp42** [mesosphere/marathon#2870]
I have a basic app running in Docker, listening on 8080 in the container, forwarded to `31022`:

```
042d3effd90f        node:5.3.0          "/bin/sh -c 'node ${M"   2 minutes ago       Up 2 minutes        0.0.0.0:31022->8080/tcp   mesos-fa72a494-eb7a-4a0b-b927-6db16019a58b-S8.e8ca0809-f3b4-4bcc-8398-07b6628b110e
```

The marathon definition has a small health check:

```json
  "healthChecks": [
    {
        "protocol": "HTTP",
        "path": "/",
        "gracePeriodSeconds": 60,
        "intervalSeconds": 10,
        "portIndex": 0,
        "timeoutSeconds": 5,
        "maxConsecutiveFailures": 200
    }
  ]
```

Unfortunately, the healthchecks are not passing as I noticed Marathon is checking the wrong IP `172.17.0.2`:

```
Dec 23 01:43:49 mesos-master-i-1208acc8 marathon[15050]: [2015-12-23 01:43:49,683] WARN Connection attempt to 172.17.0.2:31304 failed in response to GET request to / with 1 retries left, retrying... (spray.can.client.HttpHostConnectionSlot:marathon-akka.actor.default-dispatcher-768)

Dec 23 01:43:49 mesos-master-i-1208acc8 marathon[15050]: [2015-12-23 01:43:49,683] WARN Connection attempt to 172.17.0.2:31304 failed in response to GET request to / with 4 retries left, retrying... (spray.can.client.HttpHostConnectionSlot:marathon-akka.actor.default-dispatcher-768)
```

Here is an `ifconfig` on Marathon host (same on Mesos slave):

```
docker0   Link encap:Ethernet  HWaddr 02:42:ef:33:e7:5d
          inet addr:172.17.0.1  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth0      Link encap:Ethernet  HWaddr 0a:1b:a0:28:e2:63
          inet addr:10.92.42.112  Bcast:10.92.42.255  Mask:255.255.255.0
          inet6 addr: fe80::81b:a0ff:fe28:e263/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:9001  Metric:1
          RX packets:3188494 errors:0 dropped:0 overruns:0 frame:0
          TX packets:3221752 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:1165507376 (1.1 GB)  TX bytes:710361280 (710.3 MB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:2077629 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2077629 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:610406077 (610.4 MB)  TX bytes:610406077 (610.4 MB)
```

This `172.17.0.2` is the `eth0` on the actual Docker container and this is definitely wrong as it can't be reached from the actual host.

Any ideas?

**Comment by scalp42** [comment]
I see the task IP which is the docker container ip:

```json
ipAddresses: [
{
ipAddress: "172.17.0.2",
protocol: "IPv4"
}
],
ports: [
31022
],
startedAt: "2015-12-23T01:46:32.353Z",
stagedAt: "2015-12-23T01:46:30.628Z",
version: "2015-12-23T01:46:30.504Z",
slaveId: "fa72a494-eb7a-4a0b-b927-6db16019a58b-S8",
appId: "/prod/talkk/nodejs",
```

I'm not sure why the task is registered with this IP.

**Comment by scalp42** [comment]
Looks like the info comes from Mesos:

![](https://i.imgur.com/b1zI1Gq.png)

**Comment by scalp42** [comment]
I'm not sure how to override this internal ip to tell marathon to use the host ip instead.

**Comment by scalp42** [comment]
Can confirm it's a breaking change in Mesos v0.26, downgraded the cluster to v0.25 and it's now working:

![](https://i.imgur.com/CeqQ1Qk.png)

**Comment by scalp42** [comment]
See: https://issues.apache.org/jira/browse/MESOS-4113

**Comment by aquamatthias** [comment]
Thanks @scalp42 for reporting this!

**Comment by scalp42** [comment]
@aquamatthias I commented on the JIRA ticket, but do you know if there any way to add the same kind of logic that `mesos-dns` and `mesos-consul` are using ?

When we upgraded to `0.26.0`, Marathon failed **all** the health checks but our discovery mechanism survived due to these flags:

- https://github.com/CiscoCloud/mesos-consul/blob/aac6c2828a46c3a54efe1fbc41003dbcd69a6a40/main.go#L120-L123
- https://github.com/mesosphere/mesos-dns/blob/9a8aa106a05339c79fb189d435c68f64e876414c/records/config.go#L49

We had it set to `host` so that the slave ip would be picked up first no matter what.

**Comment by gkleiman** [comment]
@scalp42 thank you very much for reporting this issue!

The behaviour of using the IP address in NetworkInfo for health checks was introduced in v0.14.0-RC1, so Marathon v0.13.0 should work with Mesos v0.26.

I am working on a fix for this, which should be included in Marathon v0.14.0-RC2.

**Comment by scalp42** [comment]
@gkleiman thanks so much for looking at it!

Can you describe what is going to be the new behavior ?

If I follow the JIRA ticket correctly, Mesos 0.26 going to report the wrong Docker ip. In this case, does Marathon "discard" it and fallback to host IP like before ? I'm hoping to jump on that RC2 before our launch in 2 days (gotta live dangerously) so any insight what would need to be changed on Marathon config or tasks side would be greatly appreciated!
